lookup_plugins/asb_state.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/python
+
+DOCUMENTATION = '''
+---
+lookup: asb_state
+short_description: Retrieve state previously set by the Ansible Service Broker
+description:
+     - Provides a secure method for allowing Service Bundle developers
+       to access previously set state during a Service Bundle action.
+author:
+    - "Red Hat, Inc."
+options:
+    _terms:
+        description:
+            - Key(s) to fetch the state value for.
+        required: False
+    get_all:
+        description: >
+          Get values for all set state.
+          Returns a list with single value which is a dictionary of key, values.
+        type: bool
+        required: False
+        default: False
+    env:
+        # Set via ASB StateManager
+        - name: BUNDLE_STATE_LOCATION
+          value: /etc/apb/state
+notes: >
+  _terms as a list of keys and get_all are mutually exclusive.
+  If key(s) are provided, they will take priority and
+  their values willbe returned as a list.
+'''
+
+EXAMPLES = '''
+# Example with no keys provided and get_all arg set to True
+- name: Get all state
+  set_fact:
+    all_state: "{{lookup('asb_state', get_all=True) }}"
+
+# The following will output "The value for previously set state example=true is true"
+- debug: msg="The value for previously set state example=true is {{ all_state.example }}"
+
+# Eample for fetching a single key
+- name: Get state for single key
+  set_fact:
+    example_state: "{{lookup('asb_state', 'example') }}"
+
+# The following will output "The value for example_state state is true"
+- debug: msg="The value for example_state state is {{ example_state }}"
+
+# Example get state for multiple keys
+- name: test the new asb_state lookup plugin
+  set_fact:
+    some_test_var: "{{lookup('asb_state', 'example-key', 'example-key-two') }}"
+
+# Example to fetch state in loop
+- debug:
+    msg: 'Value is equal to {{item}}'
+  with_asb_state:
+    - 'example-key'
+    - 'example-key-two'
+'''
+
+RETURN = """
+  _raw:
+    description:
+      - The value(s) for set state
+    type: list
+"""
+
+import os
+
+from ansible.errors import AnsibleError, AnsibleFileNotFound, AnsibleParserError
+from ansible.plugins.lookup import LookupBase
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables=None, **kwargs):
+        base_state_path = os.getenv('BUNDLE_STATE_LOCATION', '/etc/apb/state')
```

Comment:
Is there any reason we're loading it from file rather than the created configmap? It seems odd to write to a configmap and read from a file, especially since it seems it would prevent you from reading state on the first provision task (not sure if that's a useful thing to do though...).

Reply:
There can only be state after the first provision. I actually can't remember how we ended up on mounting the configmap. I can go back through the PR though if needed. It may have been to allow bundle developers to read the state directly from disk if they wish, it also matches how other state is handled in pods (IE configmaps are mounted into the container rather than asking the developer to read the configmap).